pairing_api/src/uri.rs

```diff
@@ -0,0 +1,180 @@
+    chrono::Utc,
+    lazy_static::lazy_static,
+    regex::Regex,
+    serde::{Deserialize, Serialize},
+    std::collections::HashMap,
+    thiserror::Error,
+    url::Url,
+};
+
+lazy_static! {
+    static ref TOPIC_VERSION_REGEX: Regex =
+        Regex::new(r"^(?P<topic>[[:word:]-]+)@(?P<version>\d+)$").expect("Failed to compile regex");
+}
+
+#[derive(Debug, Error)]
+pub enum ParseError {
+    #[error("Invalid URI")]
+    InvalidUri,
+    #[error("Missing topic")]
+    MissingTopic,
+    #[error("Invalid version")]
+    InvalidVersion,
+    #[error("Missing symmetric key")]
+    MissingSymKey,
+    #[error("Invalid symmetric key")]
+    InvalidSymKey,
+    #[error("Missing methods")]
+    MissingMethods,
+    #[error("Invalid methods format")]
+    InvalidMethods,
+    #[error("Missing relay protocol")]
+    MissingRelayProtocol,
+    #[error("Invalid expiry timestamp")]
+    InvalidExpiryTimestamp,
+    #[error("Unexpected parameter: {0}")]
+    UnexpectedParameter(String),
+}
+
+#[derive(Debug)]
+pub struct ParsedWcUri {
+    pub topic: String,
+    pub version: String,
+    pub sym_key: String,
+    pub methods: Methods,
+    pub relay_protocol: String,
+    pub relay_data: Option<String>,
+    pub expiry_timestamp: u64,
+}
+
+pub fn parse_wc_uri(uri: &str) -> Result<ParsedWcUri, ParseError> {
+    let url = Url::parse(uri).map_err(|_| ParseError::InvalidUri)?;
+
+    if url.scheme() != "wc" {
+        return Err(ParseError::InvalidUri);
+    }
+
+    let (topic, version) = {
+        let caps = TOPIC_VERSION_REGEX
+            .captures(url.path().trim())
+            .ok_or(ParseError::InvalidUri)?;
+        let topic = caps
+            .name("topic")
+            .ok_or(ParseError::MissingTopic)?
+            .as_str()
+            .to_owned();
+        let version = caps
+            .name("version")
+            .ok_or(ParseError::InvalidVersion)?
+            .as_str()
+            .to_owned();
+
+        (topic, version)
+    };
+
+    let mut params = HashMap::new();
+    for (key, value) in url.query_pairs() {
+        params.insert(key.trim().to_string(), value.trim().to_string());
+    }
+
+    let methods_str = params.remove("methods");
+    let methods = parse_methods(methods_str.as_deref())?;
+
+    let sym_key = params.remove("symKey").ok_or(ParseError::MissingSymKey)?;
+    let relay_protocol = params
+        .remove("relay-protocol")
+        .ok_or(ParseError::MissingRelayProtocol)?;
+    let relay_data = params.remove("relay-data");
+
+    let expiry_timestamp = match params.remove("expiryTimestamp").map(|t| {
+        t.parse::<u64>()
+            .map_err(|_| ParseError::InvalidExpiryTimestamp)
+    }) {
+        None => Utc::now().timestamp() as u64 + EXPIRY_5_MINS,
+
+        Some(time) => time?,
+    };
+
+    // Check for unexpected parameters
+    if !params.is_empty() {
+        return Err(ParseError::UnexpectedParameter(
+            params.keys().next().unwrap().clone(),
+        ));
+    }
```

Comment:
I would like to avoid unwrap even if is empty false
```rust
    if let Some(unexpected_key) = params.keys().next() {
        return Err(ParseError::UnexpectedParameter(unexpected_key.clone()));
    }
```

Reply:
well yes, its safe, but if let Some pattern allows you to get unexpected parameter directly, skipping `is_empty` check

Reply:
I'll just `unwrap_or_default` since we need to return an error either ways

Reply:
Nice, resolved